File: Chapter 03/Python/section_3_3_5_dip/after/coach.py
from .player import Player

"""
DIP SOLUTION: Dependency Injection.
* ARCHITECTURE NOTE: The 'Coach' is now completely decoupled from specific 
player types. It no longer 'imports' Forward or Midfielder. 
* By accepting a list of players in the constructor, we have moved the 
responsibility of team assembly (The Composition Root) outside of this class. 
The Coach is now a pure coordinator that works with any object following 
the Player contract.
"""
class Coach:
    def __init__(self, players: list['Player']):
        # Dependency Injection: The team is provided, not created here.
        self.team = players

    def execute_game_plan(self):
        # The Coach relies on the abstraction (.perform_action()) 
        # rather than concrete methods like .attack().
        for player in self.team:
            player.perform_action()

File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_3_5_dip/after/Coach.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_3_5_dip.after;

import java.util.List;

/**
 * DIP SOLUTION: Constructor Injection and Loose Coupling.
 * * ARCHITECTURE NOTE: The Coach is no longer responsible for building its own 
 * team. The 'new' keyword has been completely removed! 
 * * Instead, the Coach asks for its dependencies through its constructor. This 
 * is called "Dependency Injection." Because the Coach only asks for a list 
 * of 'Player' abstractions, we can hand it literally any combination of players. 
 * * This makes the Coach class infinitely extensible and incredibly easy to unit 
 * test (we can just pass in "Mock" players during testing without spinning up 
 * concrete classes).
 */
public class Coach {
    // The Coach depends strictly on an abstraction!
    private final List<Player> team;

    /**
     * Dependencies are provided from the outside (Constructor Injection).
     */
    public Coach(List<Player> players) {
        this.team = players;
    }

    public void executeGamePlan() {
        for (Player player : team) {
            player.performAction();
        }
    }
}